host/ecallids.c

```diff
@@ -0,0 +1,211 @@
+    {
+        result = OE_INVALID_PARAMETER;
+        goto done;
+    }
+
+    /* Initialize output parameter to NULL id */
+    *id = OE_ECALL_ID_NULL;
+
+    /* Fetch ecall_id_table using platform agnostic API */
+    OE_CHECK(
+        oe_get_ecall_id_table(enclave, &ecall_id_table, &ecall_id_table_size));
+
+    /* Lookup/assign global id */
+    if (*global_id == OE_GLOBAL_ECALL_ID_NULL)
+    {
+        /* The global_id variable is generated by edger8r, one global id
+         *  variable per ecall. The same variable will be reused every time that
+         * ecall is made. Therefore, for an ecall, the following code will be
+         * executed only once.
+         */
+        OE_CHECK(oe_get_global_id(name, global_id));
+    }
+
+    /* Look up function id based on global id. */
+    if (*global_id >= ecall_id_table_size)
+        OE_RAISE(OE_NOT_FOUND);
+
+    /* Look-up the ecall id from the per-enclave table.
+     * The ecall_id_table is expected to be set up during enclave creation
+     */
+    *id = ecall_id_table[*global_id].id;
+    if (*id == OE_ECALL_ID_NULL)
+        OE_RAISE(OE_NOT_FOUND);
+
+    result = OE_OK;
+done:
+    return result;
+}
+
+oe_result_t oe_register_ecalls(
+    oe_enclave_t* enclave,
+    const oe_ecall_info_t* ecall_info_table,
+    uint32_t num_ecalls)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    oe_ecall_id_t* ecall_id_table = NULL;
+    uint64_t max_global_id = 0;
+    uint64_t ecall_id_table_size = 0;
+    bool locked = false;
+
+    /* Validate parameters */
+    if (!enclave || !ecall_info_table || !num_ecalls)
+    {
+        result = OE_INVALID_PARAMETER;
+        goto done;
+    }
+
+    if (oe_mutex_lock(&_lock) != 0)
+        OE_RAISE(OE_FAILURE);
+    locked = true;
+
+    /* Iterate through the ecalls and assign global ids.
+     * Also find out the maximum global id for the enclave. */
+    for (uint32_t i = 0; i < num_ecalls; i++)
+    {
+        uint64_t global_id = OE_GLOBAL_ECALL_ID_NULL;
+        const char* name = ecall_info_table[i].name;
+
+        /* Assign a proper global id based on the global __ecall_table. */
+        OE_CHECK(_get_global_id(name, &global_id));
+        if (global_id > max_global_id)
+            max_global_id = global_id;
+    }
+
+    /* Allocate ecall id table for the enclave */
+    ecall_id_table_size = max_global_id + 1;
+    ecall_id_table = malloc(sizeof(uint64_t) * ecall_id_table_size);
+    if (!ecall_id_table)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    for (uint64_t i = 0; i < ecall_id_table_size; ++i)
+        ecall_id_table[i].id = OE_ECALL_ID_NULL;
+
+    /* Fill the ecall id table */
+    for (uint32_t i = 0; i < num_ecalls; i++)
+    {
+        uint64_t global_id = OE_GLOBAL_ECALL_ID_NULL;
+        const char* name = ecall_info_table[i].name;
+        uint64_t local_id = i;
+
+        /* Assign a proper global id based on the global _ecall_table. */
+        if ((_get_global_id(name, &global_id)) != OE_OK)
```

Comment:
This seems will never be the case that returns value other than `OE_OK` now given that there is the same loop above. Just for the sake of checking, we can use `OE_CHECK` instead.

Reply:
Fixed.